### PR TITLE
Typeof can refer to block-scoped values before declaration

### LINKED ISSUE #15102: "Class XXX used before its declaration" breaking change in 2.3.0
**TypeScript Version:**  typescript@rc 2.3.0

**Code**

```ts
export interface IconProps {
    size?: number;
}

// I use this to create a type that has `size` non-nullable 
// (because it always present due to `defaultProps`
type SafeProps = IconProps & typeof Icon.defaultProps;

export default class Icon extends React.Component<IconProps, any> {
    static defaultProps = {
       size: 10
    };

    render() {
        // cast to SafeProps to safely extract `size`
        const { size } = this.props as SafeProps
        return <div style={{ size }} />
    }
}
```

**Expected behavior:**
It compiles. This actually works in 2.2.2

**Actual behavior:**

`Class 'Icon' used before its declaration` on line `type SafeProps = IconProps & typeof Icon.defaultProps;
`

Is this an intended breaking change?

### PR DESCRIPTION
**Pull request by sandersn**
Fixes #15102 by allowing references to block-scoped values before declaration inside of a type query.

Also adds `findAncestor`, which iterates up a parent pointer chain. I updated the parent pointer walks in checker to use `findAncestor`.

### CODE REVIEW

File: src/compiler/core.ts [hunk 1]
```diff
@@ -225,6 +225,28 @@ namespace ts {
...
     }
 
+    /**
+     * Iterates through the parent chain of a node and performs the callback on each parent until the callback
+     * returns a truthy value, then returns that value.
+     * If no such value is found, it applies the callback until the parent pointer is undefined or 'done' returns true
+     * At that point findAncestor returns undefined.
+     */
+    export function findAncestor<T>(node: Node, callback: (element: Node) => T | undefined, done?: (element: Node) => boolean): Node {
+        if (node) {
```
Comment (alfaproject) [inline src/compiler/core.ts]: You don't need this condition, because it's being checked in the while loop again

**Comment by sandersn** [comment]
With the PR as it stands now, total time of Monaco was 0.01 seconds faster and TFS was 0.37 seconds faster. Neither of those are significant, if I'm reading the output correctly. 

I'm going to update the PR suggestions from @alfaproject and @andy-ms and then re-run the perf test. I'll also try to get Angular correctly added to the perf test.

File: src/compiler/core.ts (hunk 1 above)
Comment (sandersn) [inline src/compiler/core.ts]: removed

**Comment by sandersn** [comment]
Perf tests are about the same with the changes suggested by @andy-ms. There's some variance but the average is right around 0.

**Pull request merged**